src/clippoly/src/clippoly.cpp

```diff
@@ -0,0 +1,271 @@
+}
+
+/**
+ * Shorthand for Boolean Union.
+ */
+Clippoly Clippoly::operator+(const Clippoly& Q) const {
+  return Boolean(Q, OpType::Add);
+}
+
+/**
+ * Shorthand for Boolean Union assignment.
+ */
+Clippoly& Clippoly::operator+=(const Clippoly& Q) {
+  *this = *this + Q;
+  return *this;
+}
+
+/**
+ * Shorthand for Boolean Difference.
+ */
+Clippoly Clippoly::operator-(const Clippoly& Q) const {
+  return Boolean(Q, OpType::Subtract);
+}
+
+/**
+ * Shorthand for Boolean Difference assignment.
+ */
+Clippoly& Clippoly::operator-=(const Clippoly& Q) {
+  *this = *this - Q;
+  return *this;
+}
+
+/**
+ * Shorthand for Boolean Intersection.
+ */
+Clippoly Clippoly::operator^(const Clippoly& Q) const {
+  return Boolean(Q, OpType::Intersect);
+}
+
+/**
+ * Shorthand for Boolean Intersection assignment.
+ */
+Clippoly& Clippoly::operator^=(const Clippoly& Q) {
+  *this = *this ^ Q;
+  return *this;
+}
+
+Clippoly Clippoly::Translate(glm::vec2 v) {
+  auto ps = C2::TranslatePaths(paths_, v.x, v.y);
+  return Clippoly(ps, true);
+}
+
+Clippoly Clippoly::Scale(glm::vec2 scale) {
+  auto scaled = C2::PathsD();
+  for (auto path : paths_) {
+    auto s = C2::PathD();
+    for (auto p : path) {
+      s.push_back(C2::PointD(p.x * scale.x, p.y * scale.y));
+    }
+    scaled.push_back(s);
+  }
+  return Clippoly(scaled, false);
+}
+
+Clippoly Clippoly::Simplify(double epsilon) {
+  auto ps = SimplifyPaths(paths_, epsilon, false);
+  return Clippoly(ps, false);
+}
+
+Clippoly Clippoly::RamerDouglasPeucker(double epsilon) {
+  auto ps = C2::RamerDouglasPeucker(paths_, epsilon);
+  return Clippoly(ps, false);
+}
+
+Clippoly Clippoly::StripNearEqual(double epsilon) {
+  auto ps = C2::StripNearEqual(paths_, epsilon, true);
+  return Clippoly(ps, false);
+}
+
+Clippoly Clippoly::StripDuplicates() {
+  auto ps = C2::StripDuplicates(paths_, true);
+  return Clippoly(ps, false);
+}
+
+Clippoly Clippoly::TrimCollinear() {
+  auto trimmed = C2::PathsD();
+  for (auto p : paths_) {
+    trimmed.push_back(C2::TrimCollinear(p, false));
+  }
+  return Clippoly(trimmed, false);
+}
+
+Clippoly Clippoly::MinkowskiSum(const std::vector<glm::vec2> pattern) {
+  auto pat = pathd_of_contour(pattern);
+  auto summed = C2::PathsD();
+  for (auto p : paths_) {
+    auto ss = C2::MinkowskiSum(pat, p, true);
+    summed.insert(summed.end(), ss.begin(), ss.end());
+  }
+  auto u = Union(summed, C2::FillRule::NonZero);
+  return Clippoly(u, true);
+}
+
+Clippoly Clippoly::MinkowskiDiff(const std::vector<glm::vec2> pattern) {
+  auto pat = pathd_of_contour(pattern);
+  auto diffed = C2::PathsD();
+  for (auto p : paths_) {
+    auto ss = C2::MinkowskiDiff(pat, p, true);
+    diffed.insert(diffed.end(), ss.begin(), ss.end());
+  }
+  auto u = Union(diffed, C2::FillRule::NonZero);
+  return Clippoly(u, true);
+}
+
+Polygons Clippoly::ToPolygons() const {
+  auto polys = Polygons();
+  auto paths =
+      clean_ ? paths_ : C2::Union(paths_, C2::FillRule::NonZero, precision_);
```

Comment:
clean_ appears to always be false? I suppose this would be for future optimization? Alternatively, we could just clean them all on construction.

Reply:
My idea with clean_ is to score a little laziness without much effort. Going through clipping sets clean_ to true, and controlled construction like square can start as clean_, while construction from arbitrary contours is unclean, since it can contain intersections. I'm a little unsure as to which non-clipping operations should be marked as unclean though.

Reply:
I think for the sake of simplicity and to match Manifold (the equivalent process is `SimplifyTopology()`), we should just clean on construction from arbitrary contours. Then we can just say all `CrossSections` are clean of self-intersections and there's no need for extra state.

Reply:
Hmm, you're right that a preemptive redundant union before other booleans probably isn't the end of the world. I still wonder if intersections can be introduced or not by the simplifications (particularly with high epsilon values), if it's possible for CrossSections to become unclean, I'd favour marking them rather than unioning every time. Not too much work to add back anyway if it does seem necessary though.